### PR TITLE
[GH-9708] Add type annotation to providers/jenkins

### PR DESCRIPTION
**Pull request by coopergillan**
Set type hints for `JenkinsHook` and `jenkins_job_trigger` ~, ignoring a couple that seem to be either slightly buggy or that I just can't quite make sense of~

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 1]
```diff
@@ -56,7 +59,7 @@ def jenkins_request_with_headers(jenkins_server, req):
         # Jenkins's funky authentication means its nigh impossible to distinguish errors.
         if e.code in [401, 403, 500]:
             raise JenkinsException(
-                'Error in request. Possibly authentication failed [%s]: %s' % (e.code, e.msg)
+                'Error in request. Possibly authentication failed [%s]: %s' % (e.code, e.msg)  # type: ignore
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]: This one threw an error about `msg` not being an attribute of `HTTPError`. How would `mypy` know that?

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 2]
```diff
@@ -160,7 +163,7 @@ def poll_job_in_queue(self, location, jenkins_server):
             location_answer = jenkins_request_with_headers(
                 jenkins_server, Request(method='POST', url=location))
             if location_answer is not None:
-                json_response = json.loads(location_answer['body'])
+                json_response = json.loads(location_answer['body'])  # type: ignore
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]: I couldn't make sense of this one:

```
incompatible type "Union[str, Mapping[str, str]]"; expected "Union[str, bytes, bytearray]"
                    json_response = json.loads(location_answer['body'])
                                               ^
```

I can't tell which one should become `str, bytes, bytearray`.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 3]
```diff
@@ -197,7 +200,7 @@ def execute(self, context):
         jenkins_server = self.get_hook().get_jenkins_server()
         jenkins_response = self.build_job(jenkins_server)
         build_number = self.poll_job_in_queue(
-            jenkins_response['headers']['Location'], jenkins_server)
+            jenkins_response['headers']['Location'], jenkins_server)  # type: ignore
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]: I can't seem to get this one either. It keeps saying this, but I can't tell which one it wants me to change:

```
airflow/providers/jenkins/operators/jenkins_job_trigger.py:204: error: Invalid index type "str" fo
"Union[str, Mapping[str, str]]"; expected type "Union[int, slice]"
                jenkins_response['headers']['Location'], jenkins_server)
                                            ^
```

**Review by coopergillan — COMMENTED** [review]
This is my first shot at doing anything with `mypy`, so some of these `ignore`s are probably not necessary. Happy to keep digging or if anyone can tell what these are easily, I would love to learn.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 4]
```diff
@@ -235,3 +238,4 @@ def execute(self, context):
             # If we can we return the url of the job
             # for later use (like retrieving an artifact)
             return build_info['url']
+        return None
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py:248]: It kinda made me add this after complaining about no return statement. It wouldn't even let me use just `return`. Is this a known issue/attribute of `mypy`?

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 1 above)
Comment (turbaszek) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Please try using `e.reason`:
```
    @property
    def reason(self):
        return self.msg
```

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 2 above)
Comment (turbaszek) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Type of `location_answer` is `Union[str, Mapping[str, str]]` and this is visible in L54:
```
return {'body': response_body.decode('utf-8'), 'headers': response_headers}
```

What I would try is:
```suggestion
                body_str: str = location_answer['body']
                json_response = json.loads(body_str)
```

The error comes from mypy assuming that the type under `body` key may be `Mapping` and that is indeed an error as `json.loads` accepts only `str, bytes, bytearray`

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 5]
```diff
@@ -19,19 +19,22 @@
...
 from airflow.exceptions import AirflowException
 from airflow.models import BaseOperator
 from airflow.providers.jenkins.hooks.jenkins import JenkinsHook
 from airflow.utils.decorators import apply_defaults
 
+JReqType = Mapping[str, Union[str, Mapping[str, str]]]
 
-def jenkins_request_with_headers(jenkins_server, req):
+
+def jenkins_request_with_headers(jenkins_server: Type[Jenkins], req: Request) -> JReqType:
```
Comment (turbaszek) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]: If the result is always in type of `{'body': ..., 'headers': ...}` I would suggest introducing named tuple. This should solve a lot of problems and will add some useful stuff like accessing using attributes.
```python
class JenkingRequest(NamedTuple):
    body: str
    headers: Dict[str, str]
```

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 1 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Awesome, that looks like it worked!

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 2 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: I couldn't quite get this to work, but managed to pull something else together. See latest commit.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 3 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Handled this one with a simpler type check.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 5 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: The headers have been a little difficult to pin down, so I went with a simpler approach.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 6]
```diff
@@ -110,7 +114,9 @@ def __init__(self,
         self.jenkins_connection_id = jenkins_connection_id
         self.max_try_before_job_appears = max_try_before_job_appears
 
-    def build_job(self, jenkins_server):
+    def build_job(
+            self, jenkins_server: Jenkins, parameters: Optional[str] = "",
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Making parameters a local variable here allow us to make the type checking here easier. It also means we are no longer updating an attribute outside of `__init__`.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 7]
```diff
@@ -195,9 +201,10 @@ def execute(self, context):
             'Triggering the job %s on the jenkins : %s with the parameters : %s',
             self.job_name, self.jenkins_connection_id, self.parameters)
         jenkins_server = self.get_hook().get_jenkins_server()
-        jenkins_response = self.build_job(jenkins_server)
-        build_number = self.poll_job_in_queue(
-            jenkins_response['headers']['Location'], jenkins_server)
+        jenkins_response = self.build_job(jenkins_server, self.parameters)
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py:207]: Here's where we preserve the same behavior for `self.build_job` by passing `self.parameters` as the "local" parameters in the function.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 8]
```diff
@@ -195,9 +201,10 @@ def execute(self, context):
...
+        if jenkins_response:
```
Comment (coopergillan) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py:208]: This is to allow for the `return None` in these responses.

**Comment by coopergillan** [comment]
@turbaszek - all right, had some good time to dig through this earlier this morning. Ended up with more changes than I expected, but am happy with where it ended up. Appreciate any other feedback!

**Comment by coopergillan** [comment]
Not sure if the Backport packages failure is due to one of my changes?

**Comment by coopergillan** [comment]
Not sure why some of these got cancelled but happy to rebase if need be.

File: airflow/providers/jenkins/hooks/jenkins.py [hunk 9]
```diff
@@ -29,7 +30,7 @@ class JenkinsHook(BaseHook):
     Hook to manage connection to jenkins server
     """
 
-    def __init__(self, conn_id='jenkins_default'):
+    def __init__(self, conn_id: Union[str, Text] = 'jenkins_default') -> None:
```
Comment (turbaszek) [inline airflow/providers/jenkins/hooks/jenkins.py]:
```suggestion
    def __init__(self, conn_id: str = 'jenkins_default') -> None:
```
That's the usual type we use for connection id

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 10]
```diff
@@ -123,20 +129,20 @@ def build_job(self, jenkins_server):
...
+        if parameters and isinstance(parameters, str):
             import ast
-            self.parameters = ast.literal_eval(self.parameters)
+            parameters = ast.literal_eval(parameters)
 
-        if not self.parameters:
-            # We need a None to call the non parametrized jenkins api end point
-            self.parameters = None
+        if not parameters:
+            # We need to a None to call the non parametrized jenkins api end point
```
Comment (turbaszek) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]:
```suggestion
            # We need None to call the non parametrized jenkins api end point
```

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 11]
```diff
@@ -123,20 +129,20 @@ def build_job(self, jenkins_server):
...
+            parameters = None
 
         request = Request(
             method='POST',
-            url=jenkins_server.build_job_url(self.job_name, self.parameters, None))
+            url=jenkins_server.build_job_url(self.job_name, parameters, None))
```
Comment (turbaszek) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]: I think at this point `parameters` have in fact `Optional[Any]` type due to `ast.literal_eval(parameters)` and this is incompatible with type `Optional[str]` of `build_job_url`.

Personally I would do:
```
params = ast.literal_eval(parameters) if parameters else None
url=jenkins_server.build_job_url(self.job_name,  params, None)
```

**Comment by turbaszek** [comment]
> Not sure if the Backport packages failure is due to one of my changes?

I restarted the jobs

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 11 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: I worked with your feedback here and what I was seeing in the code and pushed something up just now.

I think this should be ready to go now. Thanks for all your help.

**Comment by mik-laj** [comment]
We have a merge conflict. Can you do a rebase?

**Comment by coopergillan** [comment]
Yes, I can get to that soon. Thanks!

**Comment by coopergillan** [comment]
@mik-laj just rebased, now hoping for a clean run!

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 12]
```diff
@@ -94,11 +99,12 @@ class JenkinsJobTriggerOperator(BaseOperator):
...
     @apply_defaults
     def __init__(self,
-                 jenkins_connection_id,
-                 job_name,
-                 parameters="",
-                 sleep_time=10,
-                 max_try_before_job_appears=10,
+                 jenkins_connection_id: str,
+                 job_name: str,
+                 parameters: str = "",
```
Comment (mik-laj) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]:
```suggestion
                 parameters: Optional[Union[str, Dict]] = None,
```

File: airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py [hunk 13]
```diff
@@ -44,7 +44,7 @@
     job_trigger = JenkinsJobTriggerOperator(
         task_id="trigger_job",
         job_name="generate-merlin-config",
-        parameters={"first_parameter": "a_value", "second_parameter": "18"},
+        parameters='{"first_parameter": "a_value", "second_parameter": "18"}',
```
Comment (mik-laj) [inline airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: This looks like an unrelated change.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 4 above)
Comment (mik-laj) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py:248]: This is mypa's expected behavior. If the function exits without calling `return`, Python implicitly returns None. Mypy requires us to be implicit in this case.

 > Explicit is better than implicit.

https://www.python.org/dev/peps/pep-0020/

File: airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py (hunk 13 above)
Comment (potiuk) [reply airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: I think it's because of the wrong type specified in the docstring (and yes it should not be changed here)

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 12 above)
Comment (potiuk) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Actually, when you look at the "build_job_url" method code and docstring, it should be Optional[Union[str, Dict, List]]

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 14]
```diff
@@ -56,7 +60,7 @@ def jenkins_request_with_headers(jenkins_server, req):
         # Jenkins's funky authentication means its nigh impossible to distinguish errors.
         if e.code in [401, 403, 500]:
             raise JenkinsException(
-                'Error in request. Possibly authentication failed [%s]: %s' % (e.code, e.msg)
+                'Error in request. Possibly authentication failed [%s]: %s' % (e.code, e.reason)
```
Comment (potiuk) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py:64]: Nice!

**Review by potiuk — CHANGES_REQUESTED** [review]

File: airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py (hunk 13 above)
Comment (coopergillan) [reply airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: I changed this since I was seeing that `parameters` was a templated string. Now we have determined that it can be a JSON blob string, a path to a JSON file, a dictionary, or a list.

It seems like only the dictionary case is tested here. I'm thinking there should be test coverage for each of these cases, right?

If so, I can open a GH Issue. What do you think @mik-laj @potiuk?

Comment (coopergillan) [reply airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: See comments after latest commit.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 12 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: See latest update.

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py (hunk 4 above)
Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py:248]: Got it, that makes sense :+1:

File: airflow/providers/jenkins/operators/jenkins_job_trigger.py [hunk 15]
```diff
@@ -80,7 +86,7 @@ class JenkinsJobTriggerOperator(BaseOperator):
     :param job_name: The name of the job to trigger
     :type job_name: str
     :param parameters: The parameters block to provide to jenkins. (templated)
-    :type parameters: str
+    :type parameters: JSON blob, path to JSON file, Dict, or List
```
Comment (turbaszek) [inline airflow/providers/jenkins/operators/jenkins_job_trigger.py]:
```
    :param parameters: The parameters block to provide to jenkins. JSON blob, path to JSON file, Dict, or List. (templated)
    :type parameters: Union[str, Dict, List]
```
I think the type should be explicit, WDYT?

Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: You mean essentially just put the exact type like you have here? That seems like wider practice, right? I can make that change.

Comment (turbaszek) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: Yes, and you can add the information about json etc to `:param` part so users understand what the type mean

Comment (coopergillan) [reply airflow/providers/jenkins/operators/jenkins_job_trigger.py]: @turbaszek - gave this a shot in the latest commit. Let me know what you think.

**Comment by coopergillan** [comment]
I picked Jenkins for my first typing PR since it looked small. Turns out there is a lot going on here!

**Review by turbaszek — APPROVED** [review]
LGTM!

**Comment by turbaszek** [comment]
> I picked Jenkins for my first typing PR since it looked small. Turns out there is a lot going on here!

Hah, type annotations can be tricky sometimes but I think it's always an interesting lesson 👌

**Comment by coopergillan** [comment]
@turbaszek - any thoughts on the question about improving the unit tests? I asked the others a question here: https://github.com/apache/airflow/pull/9947#discussion_r464084255

File: airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py (hunk 13 above)
Comment (turbaszek) [reply airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: > It seems like only the dictionary case is tested here. I'm thinking there should be test coverage for each of these cases, right?

It would be nice to use parametrize to test each of possible type

**Comment by coopergillan** [comment]
Just need to know if @potiuk is good with this one now.

File: airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py (hunk 13 above)
Comment (coopergillan) [reply airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: Yeah, I was thinking something like that. I will raise an issue.

Comment (coopergillan) [reply airflow/providers/jenkins/example_dags/example_jenkins_job_trigger.py]: https://github.com/apache/airflow/issues/10132

**Comment by coopergillan** [comment]
@potiuk - this one should be ready to go. Last one from me with the GH- commit messages 😅

**Comment by potiuk** [comment]
Can you rebase on top of master please :)

**Comment by coopergillan** [comment]
Done with a merge in.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Thanks!
